src/Interpreters/ProcessList.cpp

```diff
@@ -430,12 +437,37 @@ QueryStatus * ProcessList::tryGetProcessListElement(const String & current_query
 
 CancellationCode ProcessList::sendCancelToQuery(const String & current_query_id, const String & current_user, bool kill)
 {
-    auto lock = safeLock();
+    QueryStatus * elem;
+    CancelledProcesses::Iterator cancelled_it;
+
+    /// Cancelling the query should be done without the lock.
```

Comment:
Similar fix is required for `killAllQueries` to fix https://github.com/ClickHouse/ClickHouse/issues/41001 completely